spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -16,191 +16,111 @@
+    val withKeys: RDD[(Iterator[SpatialKey], Feature[G,D])] =
+      rdd.map(f => (mapTransform.keysForGeometry(f.geom), f))
+
+    /* Clip every Feature */
+    withKeys.flatMap { case (ks, f) => ks.flatMap(k => clipFeature(k.extent(layout), f).map((k, _))) }
+  }
+
+  /** Clips Features to fit the given Extent. Does its best to avoid unnecessary
+    * JTS interactions.
+    */
+  def byExtent[G <: Geometry, D](extent: Extent, f: Feature[G, D]): Option[Feature[Geometry, D]] = f.geom match {
+    case g: Point                              => Some(f)
+    case g: MultiPoint if coveredBy(g, extent) => Some(f)
+    case g: MultiPoint                         => clip(extent, g).map(Feature(_, f.data))
+    case g: Line       if coveredBy(g, extent) => Some(f)
+    case g: Line                               => clip(extent, g).map(Feature(_, f.data))
+    case g: MultiLine  if coveredBy(g, extent) => Some(f)
+    case g: MultiLine                          => clip(extent, g).map(Feature(_, f.data))
+    case g =>  /* Polygon and MultiPolygon */
+      val pg = PreparedGeometryFactory.prepare(g.jtsGeom)
+      val ep = extent.toPolygon.jtsGeom
+
+      if (pg.covers(ep)) Some(Feature(extent, f.data))
+      else if (pg.coveredBy(ep)) Some(f)
+      else clip(extent, g).map(Feature(_, f.data))
   }
 
-  private def clipGeom[G <: Geometry, D](
-    mapTransform: MapKeyTransform,
-    clipFeature: (Extent, Feature[G, D], Predicates) => Option[Feature[Geometry, D]],
-    feature: Feature[G, D]
-  ): Iterator[(SpatialKey, Feature[Geometry, D])] = {
-    val pointPredicates =
-      new Predicates {
-        def covers(e: Extent) = false
-        def coveredBy(e: Extent) = true
-      }
-
-    val mpOrLinePredicates =
-      new Predicates {
-        def covers(e: Extent) = false
-        def coveredBy(e: Extent) =
-          feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def polyPredicates(pg: PreparedGeometry) =
-      new Predicates {
-        def covers(e: Extent) = pg.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = pg.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def gcPredicates =
-      new Predicates {
-        def covers(e: Extent) = feature.geom.jtsGeom.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def clipToKey(
-      k: SpatialKey,
-      preds: Predicates
-    ): Option[(SpatialKey, Feature[Geometry, D])] = {
-      val extent = mapTransform(k)
-
-      clipFeature(extent, feature, preds).map(k -> _)
+  private def coveredBy[G <: Geometry](g: G, e: Extent): Boolean =
+    g.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
+
+  private def clip[G <: Geometry](e: Extent, g: G): Option[Geometry] = {
+    val exPoly: Polygon = e.toPolygon
+
+    val clipped: Try[Geometry] = g match {
+      case mp: MultiPolygon => Try(MultiPolygon(mp.polygons.flatMap(_.intersection(exPoly).as[Polygon])))
```

Comment:
Not sure if this is the best thing to do. It was a work-around from long ago to help prevent clips on large MultiPolygons frmo going insane.

Reply:
No, we should let the underlying intersection logic of JTS's MultiPolygon do it's thing.

Reply:
Cool, I can remove it. It was a fairly old work-around.